src/boat_simulator/boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,260 @@
+
+        Args:
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in degrees/second.
+            `max_angle_range` (tuple): Tuple of minimum to maximum control angle range
+        """
+        self.current_control_ang = current_control_ang
+        self.time_step = time_step
+        self.max_angle_range = max_angle_range
+
+    @abstractmethod
+    def update_state(self) -> bool:
+        """Updates the controller position based on target heading angles
+
+        Returns:
+            False if target angle was not reached, else True if desired heading has been reached
+        """
+
+        pass
+
+    @abstractmethod
+    def compute_error_angle(self) -> Scalar:
+        """Computes the corresponding feedback based on error and desired heading"""
+        pass
+
+
+class RudderController(ActuatorController):
+    """General Class for the Actuator Controller.
+
+    RudderController Extends: ActuatorController
+    """
+
+    def __init__(
+        self,
+        current_heading: Scalar,
+        desired_heading: Scalar,
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        kp: Scalar,
+        cp: Scalar,
+        max_angle_range: tuple,
+        rudder_speed: Scalar,
+    ):
+        """Initializes the class attributes.
+
+        Args:
+            `current_heading` (Scalar): Current boat heading direction in degrees.
+            `desired_heading` (Scalar): Target boating heading direction in degrees.
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in degrees/second.
+            `kp` (Scalar): Proportional constant when calculating error.
+            `cp` (Scalar): Timeseries of error values computed over time.
+            `max_angle_range` (tuple): Tuple of minimum to maximum control angle range in degrees
+            `rudder_speed` (Scalar): Speed of controller change in degrees per second
+
+        """
+        super().__init__(current_control_ang, time_step, max_angle_range)
+
+        self.current_heading = current_heading
+        self.desired_heading = desired_heading
+        self.kp = kp
+        self.cp = cp
+        self.max_angle_range = max_angle_range
+        self.rudder_speed = rudder_speed
+        self.running_error = 0.0  # keeps track of remaining error between desired and current
+
+    def compute_error(
+        self, desired: Scalar, current: Scalar
+    ) -> Scalar:  # angle passed in as radians
+        """Computes the error between desired and current heading
+        implementation taken from: https://stackoverflow.com/a/2007279
+        Angles are bound with the convention (-180, 180]
+
+        Args:
+            `desired` (Scalar): New desired heading in degrees
+            `current` (Scalar): New current heading in degrees
```

Comment:
I will need to change this convention. I believe I'm bounding (-180, 180], so the zero is currently at the positive x. I'll alter my parameters to take in this convention

Reply:
For this can I just specify when the user inputs an angle to use this convention? This way when they pass in the degree I can still bound it accordingly, or am I simplifying it too much?

Reply:
The desired heading and current heading will both be sensor readings that you'll get directly from the ROS network, so they are already in that convention. It's a matter of convenience for you to adopt this convention in your actuator to avoid switching conventions.

Reply:
So yes, you want to tell the use what convention to use. Euler angles are unfortunately arbitrary sometimes, so we are forced to specify angle conventions all the time to avoid confusion.

Reply:
Sounds good. If my understanding is correct, there wasn't anything in my code that really depended on the convention; it would just be important for the user to understand what convention is being used when understanding how to utilize my controller right? I made changes to specify the angle convention in the doc string.